### PR TITLE
#21 One little step

### PR DESCRIPTION
**Pull request by nenros**
Prepared everything for next step, basic genserver created, added pid to both loops.
I don't have idea for tests for know, loops breaks everything.

### CODE REVIEW

File: lib/zmq_ex.ex [hunk 1]
```diff
@@ -1,57 +1,84 @@
...
-    {:ok, socket} = :gen_tcp.accept(listen_socket)
-
-    :gen_tcp.send(socket, "connected!")
+  @doc """
+  Start server or client on given port.
+  Returns pid of socket process.
 
+  """
+  @spec start(type :: :server | :client, port :: integer) :: pid
+  def start(type, port) do
```
Comment (fabiankrol) [inline lib/zmq_ex.ex]: Move implementation inside `start_link`. `start` can be removed

File: lib/zmq_ex.ex [hunk 2]
```diff
@@ -1,57 +1,84 @@
...
+    Logger.debug(fn -> "Starting #{type} on port:#{port}" end)
+
+    {:ok, socket} =
```
Comment (fabiankrol) [inline lib/zmq_ex.ex]: Don't create socket in here, move it inside gen_server's `init` callback

File: lib/zmq_ex.ex [hunk 3]
```diff
@@ -1,57 +1,84 @@
...
-  def start_client do
-    opts = [:binary, active: false]
-    {:ok, socket} = :gen_tcp.connect('localhost', 5555, opts)
+  def start_link(socket) do
+    GenServer.start_link(__MODULE__, socket)
+  end
 
-    :gen_tcp.send(socket, "connected!")
+  @impl true
+  def init(socket) do
```
Comment (fabiankrol) [inline lib/zmq_ex.ex]: for long initialization use ["A similar approach"](http://blog.lucidsimple.com/2016/02/07/simple-OTP-idioms-using-handle-info-part-1.html).

Comment (fabiankrol) [reply lib/zmq_ex.ex]: cc @Dreat

File: lib/zmq_ex.ex [hunk 4]
```diff
@@ -1,57 +1,84 @@
 defmodule ZmqEx do
+  use GenServer
+
   @moduledoc """
   Documentation for ZmqEx.
   """
 
+  @socket_opts [:binary, active: false]
```
Comment (fabiankrol) [inline lib/zmq_ex.ex]: Investigate other `active` configuration options to implement `rec_loop` using gen_server callback

File: lib/zmq_ex.ex (hunk 3 above)
Comment (Dreat) [reply lib/zmq_ex.ex]: +1 for similar approach

File: lib/zmq_ex.ex (hunk 4 above)
Comment (Dreat) [reply lib/zmq_ex.ex]: Do we need it right now? Maybe we can skip it for later, I think this can be out of scope for now

Comment (nenros) [reply lib/zmq_ex.ex]: changed to `true`, added message handler, I left active  `false` for start connection to handle old solution.

File: lib/zmq_ex.ex (hunk 3 above)
Comment (nenros) [reply lib/zmq_ex.ex]: is it ok now?

File: lib/zmq_ex.ex (hunk 2 above)
Comment (nenros) [reply lib/zmq_ex.ex]: moved there

File: lib/zmq_ex.ex (hunk 1 above)
Comment (nenros) [reply lib/zmq_ex.ex]: as you wish ;)

**Comment by fabiankrol** [comment]
Let's continue this on local feature branch: #30

**Pull request merged**